src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
+		// since no valid velocity, update derivate with 0
+		_vel_x_deriv.update(0.0f);
+		_vel_y_deriv.update(0.0f);
 	}
 
-	/* angle = cos(x) + 1.0
-	 * angle goes from 0 to 2 with 0 = large angle, 2 = small angle:   0 = PI ; 2 = PI*0 */
-	float angle = 2.0f;
-
-	if (unit_current_to_next.length() > SIGMA_NORM) {
-		angle = unit_current_to_next * (unit_prev_to_current * -1.0f) + 1.0f;
-	}
+	if (MPC_ALT_MODE.get() && _local_pos.dist_bottom_valid && PX4_ISFINITE(_local_pos.dist_bottom_rate)) {
+		// terrain following
+		_states.velocity(2) = -_local_pos.dist_bottom_rate;
+		_states.acceleration(2) = _vel_z_deriv.update(-_states.velocity(2));
 
-	/* compute velocity target close to waypoint */
-	float vel_close;
+	} else if (PX4_ISFINITE(_local_pos.vz)) {
 
-	if (use_linear_approach) {
+		if (PX4_ISFINITE(vel_sp_z) && fabsf(vel_sp_z) > FLT_EPSILON && PX4_ISFINITE(_local_pos.z_deriv)) {
+			// A change in velocity is demanded. Set velocity to the derivative of position
+			// because it has less bias but blend it in across the landing speed range
+			float weighting = fminf(fabsf(vel_sp_z) / _land_speed.get(), 1.0f);
+			_states.velocity(2) = _local_pos.z_deriv * weighting + _local_pos.vz * (1.0f - weighting);
+		}
 
-		/* velocity close to target adjusted to angle
-		 * vel_close =  m*x+q
-		 */
-		float slope = -(get_cruising_speed_xy() - min_cruise_speed) / 2.0f;
-		vel_close = slope * angle + get_cruising_speed_xy();
+		_states.velocity(2) = _local_pos.vz;
+		_states.acceleration(2) = _vel_z_deriv.update(-_states.velocity(2));
 
 	} else {
+		_states.velocity(2) = _states.acceleration(2) = NAN;
+		// since no valid velocity, update derivate with 0
+		_vel_z_deriv.update(0.0f);
 
-		/* velocity close to target adjusted to angle
-		 * vel_close = a *b ^x + c; where at angle = 0 -> vel_close = vel_cruise; angle = 1 -> vel_close = middle_cruise_speed (this means that at 90degrees
-		 * the velocity at target is middle_cruise_speed);
-		 * angle = 2 -> vel_close = min_cruising_speed */
-
-		/* from maximum cruise speed, minimum cruise speed and middle cruise speed compute constants a, b and c */
-		float a = -((middle_cruise_speed -  get_cruising_speed_xy()) * (middle_cruise_speed -  get_cruising_speed_xy())) /
-			  (2.0f * middle_cruise_speed - get_cruising_speed_xy() - min_cruise_speed);
-		float c =  get_cruising_speed_xy() - a;
-		float b = (middle_cruise_speed - c) / a;
-		vel_close = a * powf(b, angle) + c;
 	}
 
-	/* vel_close needs to be in between max and min */
-	return math::constrain(vel_close, min_cruise_speed, get_cruising_speed_xy());
-
-}
-
-float
-MulticopterPositionControl::get_cruising_speed_xy()
-{
-	/*
-	 * in mission the user can choose cruising speed different to default
-	 */
-	return ((PX4_ISFINITE(_pos_sp_triplet.current.cruising_speed) && !(_pos_sp_triplet.current.cruising_speed < 0.0f)) ?
-		_pos_sp_triplet.current.cruising_speed : _vel_cruise_xy.get());
+	if (PX4_ISFINITE(_local_pos.yaw)) {
+		_states.yaw = _local_pos.yaw;
```

Comment:
@MaEtUgR @Stifael We need to check is not isfinite implies that it is NaN. Isn't ther also stuff like INFINITY?

Reply:
INFINITY and NAN should both return 0 is checked with PX4_ISFINITE.